In [137]:
from pathlib import Path

import numpy as np
import pandas as pd

In [168]:
MARCH_RAW_DIR = Path("../data/raw/pcdd/2026-03")
JUNE_RAW_DIR = Path("../data/raw/pcdd/2026-06")

june_paths = {
    "nhs_rate": JUNE_RAW_DIR / "pcdem-nhs-rate-jun-2026.csv",
    "la_rate": JUNE_RAW_DIR / "pcdem-la-rate-jun-2026.csv",
    "sub_icb": JUNE_RAW_DIR / "pcdem-sub-icb-jun-2026.csv",
    "practice": JUNE_RAW_DIR / "pcdem-practice-jun-2026.csv",
    "mapping": JUNE_RAW_DIR / "mapping-file-july-dementia-2026.csv",
    "dictionary": JUNE_RAW_DIR / "PCDD-2627-data-dictionary.xlsx",
}

nhs_rate = pd.read_csv(june_paths["nhs_rate"])
la_rate = pd.read_csv(june_paths["la_rate"])
sub_icb = pd.read_csv(june_paths["sub_icb"])
practice = pd.read_csv(june_paths["practice"])
mapping = pd.read_csv(june_paths["mapping"])
dictionary = pd.read_excel(june_paths["dictionary"])

datasets = {"nhs_rate": nhs_rate, "la_rate": la_rate, "sub_icb": sub_icb, "practice": practice, "mapping": mapping}

for name, dataset in datasets.items():
    print(f"Displaying dataset: {name}")
    display(dataset)
    print(f"shape of {name}: {dataset.shape}")
    print()

Displaying dataset: nhs_rate


,INDICATOR,ORG_TYPE,ORG_CODE,ONS_CODE,NAME,ACH_DATE,MEASURE,VALUE,DQ
0,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_RESPONSIBILITY,ENG,ENG,ENGLAND,30-Jun-26,DEMENTIA_ESTIMATE_65_PLUS,756824.7,NaN
1,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_RESPONSIBILITY,ENG,ENG,ENGLAND,30-Jun-26,DEMENTIA_REGISTER_65_PLUS,501216.0,NaN
2,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_RESPONSIBILITY,ENG,ENG,ENGLAND,30-Jun-26,DIAG_RATE_65_PLUS,66.2,NaN
3,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_RESPONSIBILITY,ENG,ENG,ENGLAND,30-Jun-26,DIAG_RATE_65_PLUS_LL,59.6,NaN
4,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_RESPONSIBILITY,ENG,ENG,ENGLAND,30-Jun-26,DIAG_RATE_65_PLUS_UL,71.6,NaN
...,...,...,...,...,...,...,...,...,...
745,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,SUB_ICB_LOC,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,30-Jun-26,DEMENTIA_ESTIMATE_65_PLUS,5172.2,NaN
746,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,SUB_ICB_LOC,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,30-Jun-26,DEMENTIA_REGISTER_65_PLUS,3431.0,NaN
747,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,SUB_ICB_LOC,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,30-Jun-26,DIAG_RATE_65_PLUS,66.3,NaN
748,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,SUB_ICB_LOC,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,30-Jun-26,DIAG_RATE_65_PLUS_LL,59.3,NaN


shape of nhs_rate: (750, 9)

Displaying dataset: la_rate


,INDICATOR,ORG_TYPE,ONS_CODE,NAME,ACH_DATE,MEASURE,VALUE,DQ
0,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_GEOGRAPHICAL,E92000001,ENGLAND,30-Jun-26,DEMENTIA_ESTIMATE_65_PLUS,756824.7,NaN
1,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_GEOGRAPHICAL,E92000001,ENGLAND,30-Jun-26,DEMENTIA_REGISTER_65_PLUS,501216.0,NaN
2,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_GEOGRAPHICAL,E92000001,ENGLAND,30-Jun-26,DIAG_RATE_65_PLUS,66.2,NaN
3,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_GEOGRAPHICAL,E92000001,ENGLAND,30-Jun-26,DIAG_RATE_65_PLUS_LL,59.6,NaN
4,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,COUNTRY_GEOGRAPHICAL,E92000001,ENGLAND,30-Jun-26,DIAG_RATE_65_PLUS_UL,71.6,NaN
...,...,...,...,...,...,...,...,...
2290,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,UTLA,E10000034,Worcestershire,30-Jun-26,DEMENTIA_ESTIMATE_65_PLUS,9956.3,NaN
2291,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,UTLA,E10000034,Worcestershire,30-Jun-26,DEMENTIA_REGISTER_65_PLUS,5548.0,NaN
2292,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,UTLA,E10000034,Worcestershire,30-Jun-26,DIAG_RATE_65_PLUS,55.7,NaN
2293,DEMENTIA: 65+ ESTIMATED DIAGNOSIS RATE,UTLA,E10000034,Worcestershire,30-Jun-26,DIAG_RATE_65_PLUS_LL,50.1,NaN


shape of la_rate: (2295, 8)

Displaying dataset: sub_icb


,ACH_DATE,ORG_TYPE,ODS_CODE,ONS_CODE,NAME,MEASURE,BREAKDOWN,AGE,GENDER,ETHNICITY,DEMENTIA_TYPE,RESIDENTIAL_TYPE,VALUE
0,30-Jun-26,SUB_ICB,00L,E38000130,NHS NORTH EAST AND NORTH CUMBRIA ICB - 00L,DELIRIUM_12M,NaN,ALL,ALL,ALL,ALL,ALL,262
1,30-Jun-26,SUB_ICB,00L,E38000130,NHS NORTH EAST AND NORTH CUMBRIA ICB - 00L,DEMENTIA_REGISTER,AGE_GENDER,0_39,Female,ALL,ALL,ALL,2
2,30-Jun-26,SUB_ICB,00L,E38000130,NHS NORTH EAST AND NORTH CUMBRIA ICB - 00L,DEMENTIA_REGISTER,AGE_GENDER,40_44,Female,ALL,ALL,ALL,0
3,30-Jun-26,SUB_ICB,00L,E38000130,NHS NORTH EAST AND NORTH CUMBRIA ICB - 00L,DEMENTIA_REGISTER,AGE_GENDER,45_49,Female,ALL,ALL,ALL,2
4,30-Jun-26,SUB_ICB,00L,E38000130,NHS NORTH EAST AND NORTH CUMBRIA ICB - 00L,DEMENTIA_REGISTER,AGE_GENDER,50_54,Female,ALL,ALL,ALL,3
...,...,...,...,...,...,...,...,...,...,...,...,...,...
11655,30-Jun-26,SUB_ICB,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,PRESCRIBING,PSY_DIAG,ALL,ALL,ALL,ALL,ALL,172
11656,30-Jun-26,SUB_ICB,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,PRESCRIBING,PSY_DIAG_WITH_ANTIPSYMEDS,ALL,ALL,ALL,ALL,ALL,104
11657,30-Jun-26,SUB_ICB,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,REFERRALS,NUM_PAT_REF_MEM_CLIN_12M,ALL,ALL,ALL,ALL,ALL,718
11658,30-Jun-26,SUB_ICB,X2C4Y,E38000254,NHS WEST YORKSHIRE ICB - X2C4Y,REFERRALS,NUM_REF_MEM_CLIN_12M,ALL,ALL,ALL,ALL,ALL,849


shape of sub_icb: (11660, 13)

Displaying dataset: practice


,ACH_DATE,ORG_TYPE,ODS_CODE,NAME,MEASURE,BREAKDOWN,VALUE
0,30-Jun-26,PRACTICE,A81001,THE DENSHAM SURGERY,DEMENTIA_REGISTER,DEMENTIA_REGISTER_0_64,0
1,30-Jun-26,PRACTICE,A81001,THE DENSHAM SURGERY,DEMENTIA_REGISTER,DEMENTIA_REGISTER_65_PLUS,28
2,30-Jun-26,PRACTICE,A81001,THE DENSHAM SURGERY,PAT_LIST,PAT_LIST_0_64,2850
3,30-Jun-26,PRACTICE,A81001,THE DENSHAM SURGERY,PAT_LIST,PAT_LIST_65_PLUS,883
4,30-Jun-26,PRACTICE,A81001,THE DENSHAM SURGERY,REVIEWS,DIAG_DECLINED_CARE_PLAN,0
...,...,...,...,...,...,...,...
42611,30-Jun-26,PRACTICE,Y08411,ONE CARE LEWISHAM,PAT_LIST,PAT_LIST_0_64,38
42612,30-Jun-26,PRACTICE,Y08411,ONE CARE LEWISHAM,PAT_LIST,PAT_LIST_65_PLUS,288
42613,30-Jun-26,PRACTICE,Y08411,ONE CARE LEWISHAM,REVIEWS,DIAG_DECLINED_CARE_PLAN,0
42614,30-Jun-26,PRACTICE,Y08411,ONE CARE LEWISHAM,REVIEWS,DIAG_RECEIVED_CARE_PLAN,17


shape of practice: (42616, 7)

Displaying dataset: mapping


,EXTRACT_DATE,PRACTICE_CODE,PRACTICE_NAME,PRACTICE_POSTCODE,ONS_SUB_ICB_LOCATION_CODE,SUB_ICB_LOCATION_CODE,SUB_ICB_LOCATION_NAME,ONS_ICB_CODE,ICB_CODE,ICB_NAME,ONS_COMM_REGION_CODE,COMM_REGION_CODE,COMM_REGION_NAME
0,01-Jul-26,A81001,THE DENSHAM SURGERY,TS18 1HU,E38000247,16C,NHS NORTH EAST AND NORTH CUMBRIA ICB - 16C,E54000050,QHM,NHS NORTH EAST AND NORTH CUMBRIA INTEGRATED CA...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
1,01-Jul-26,A81002,QUEENS PARK MEDICAL CENTRE,TS18 2AW,E38000247,16C,NHS NORTH EAST AND NORTH CUMBRIA ICB - 16C,E54000050,QHM,NHS NORTH EAST AND NORTH CUMBRIA INTEGRATED CA...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
2,01-Jul-26,A81004,ACKLAM MEDICAL CENTRE,TS5 8SB,E38000247,16C,NHS NORTH EAST AND NORTH CUMBRIA ICB - 16C,E54000050,QHM,NHS NORTH EAST AND NORTH CUMBRIA INTEGRATED CA...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
3,01-Jul-26,A81005,SPRINGWOOD SURGERY,TS14 7DJ,E38000247,16C,NHS NORTH EAST AND NORTH CUMBRIA ICB - 16C,E54000050,QHM,NHS NORTH EAST AND NORTH CUMBRIA INTEGRATED CA...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
4,01-Jul-26,A81006,TENNANT STREET MEDICAL PRACTICE,TS18 2AT,E38000247,16C,NHS NORTH EAST AND NORTH CUMBRIA ICB - 16C,E54000050,QHM,NHS NORTH EAST AND NORTH CUMBRIA INTEGRATED CA...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
...,...,...,...,...,...,...,...,...,...,...,...,...,...
6177,01-Jul-26,Y08411,ONE CARE LEWISHAM,SE23 1RJ,E38000244,72Q,NHS SOUTH EAST LONDON ICB - 72Q,E54000030,QKK,NHS SOUTH EAST LONDON INTEGRATED CARE BOARD,E40000003,Y56,LONDON COMMISSIONING REGION
6178,01-Jul-26,Y08516,YORK GP EXTENDED ACCESS (NIMBUSCARE),YO24 4LJ,E38000188,03Q,NHS HUMBER AND NORTH YORKSHIRE ICB - 03Q,E54000051,QOQ,NHS HUMBER AND NORTH YORKSHIRE INTEGRATED CARE...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
6179,01-Jul-26,Y08645,WAVERLEY MEDICAL CENTRE,S60 8GD,E38000141,03L,NHS SOUTH YORKSHIRE ICB - 03L,E54000061,QF7,NHS SOUTH YORKSHIRE INTEGRATED CARE BOARD,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION
6180,01-Jul-26,Y08852,STAYCITY YORK (NIMBUSCARE),YO24 4LJ,E38000188,03Q,NHS HUMBER AND NORTH YORKSHIRE ICB - 03Q,E54000051,QOQ,NHS HUMBER AND NORTH YORKSHIRE INTEGRATED CARE...,E40000012,Y63,NORTH EAST AND YORKSHIRE COMMISSIONING REGION


shape of mapping: (6182, 13)



In [227]:
xls = pd.ExcelFile(june_paths["dictionary"])
display(xls.sheet_names)

['Title Sheet',
 'Notes and caveats',
 'pcdem-nhs-rate',
 'pcdem-la-rate',
 'pcdem-practice',
 'pcdem-sub-icb',
 'Age_Gender categories',
 'Ethnicity categories',
 'Dementia Type categories',
 'Residential Type categories',
 'mapping-file-mmm-dementia-yyyy']

## 2. Dataset structure and grain

Before analysing values, establish what one row represents in each dataset and whic columns uniquely indentify a record.

In [200]:
for name, df in datasets.items():
    print(f"\n{name.upper()}")
    print((f"Rows: {len(df):,}"))
    print(f"Columns: {len(df.columns)}")
    print("Columns:")
    print(df.columns.tolist())
    


NHS_RATE
Rows: 750
Columns: 9
Columns:
['INDICATOR', 'ORG_TYPE', 'ORG_CODE', 'ONS_CODE', 'NAME', 'ACH_DATE', 'MEASURE', 'VALUE', 'DQ']

LA_RATE
Rows: 2,295
Columns: 8
Columns:
['INDICATOR', 'ORG_TYPE', 'ONS_CODE', 'NAME', 'ACH_DATE', 'MEASURE', 'VALUE', 'DQ']

SUB_ICB
Rows: 11,660
Columns: 13
Columns:
['ACH_DATE', 'ORG_TYPE', 'ODS_CODE', 'ONS_CODE', 'NAME', 'MEASURE', 'BREAKDOWN', 'AGE', 'GENDER', 'ETHNICITY', 'DEMENTIA_TYPE', 'RESIDENTIAL_TYPE', 'VALUE']

PRACTICE
Rows: 42,616
Columns: 7
Columns:
['ACH_DATE', 'ORG_TYPE', 'ODS_CODE', 'NAME', 'MEASURE', 'BREAKDOWN', 'VALUE']

MAPPING
Rows: 6,182
Columns: 13
Columns:
['EXTRACT_DATE', 'PRACTICE_CODE', 'PRACTICE_NAME', 'PRACTICE_POSTCODE', 'ONS_SUB_ICB_LOCATION_CODE', 'SUB_ICB_LOCATION_CODE', 'SUB_ICB_LOCATION_NAME', 'ONS_ICB_CODE', 'ICB_CODE', 'ICB_NAME', 'ONS_COMM_REGION_CODE', 'COMM_REGION_CODE', 'COMM_REGION_NAME']


In [201]:
nhs_rate["DQ"].isna().mean()
nhs_rate["DQ"].value_counts(dropna=False)

DQ
NaN    750
Name: count, dtype: int64

## Rate Dataset

**Publisher definition:** Diagnosis rate indicator by sub ICB location - ICB - region - England

**Shape:** 750 rows × 9 columns

### What does one row represent?

The rate dataset reports dementia measures for NHS organisations/georgaphies at the reporting period end date.

Key fields:
- `NAME` - organisation name
    - an organisation is a boundary definition such as 'ENGLAND' or 'NHS NORTH WEST YORKSHIRE ICB - X2C4Y
- `ACH_DATE` - achieved date
    - the achieved date represnts the reporting period end date
    - this terminology is inherited from the Quality and Outcomes Framework (QOF) system
- `MEASURE` - identifies what the `VALUE` represents
- `VALUE` - alue associated with the measure
- `ORG_TYPE` - organisation breakdown (geographic level)
- `ORG_CODE` - ODS (Organisation Data Service) organisation identifier
- `ONS_CODE` - 9-digit geography code of the organisation
- `DQ` - "Identifies organisations with a smaller denominator population than the CFAS II reference population with a ‘1’. Values for these organisations should be interpreted with caution."

Key measures:
|Measure|Meaning|
|---|---|
| DEMENTIA_REGISTER_65_PLUS | Number of patients aged 65 and over with a recorded dementia diagnosis |
| DEMENTIA_ESTIMATE_65_PLUS | Expected number of patients aged 65 and over with dementia |
| DIAG_RATE_65_PLUS | Estimated dementia diagnosis rate (recorded/expected) in patients aged 65 and over |
| DIAG_RATE_65_PLUS_LL | Lower 95% confidence limit of estimated dementia diagnosis rate in patients aged 65 and over |
| DIAG_RATE_65_PLUS_UL | Upper 95% confidence limit of estimated dementia diagnosis rate in patients aged 65 and over |



### ORG_TYPE

In [202]:
# Count rate observations by geographic / organisation level.
nhs_rate["ORG_TYPE"].value_counts().to_frame()


,count
ORG_TYPE,
SUB_ICB_LOC,530
ICB,180
NHS_REGION,35
COUNTRY_RESPONSIBILITY,5


In [203]:
# Count the unique reporting entities at each geographic / organisation level.
nhs_rate.groupby("ORG_TYPE")["ORG_CODE"].nunique().to_frame()


,ORG_CODE
ORG_TYPE,
COUNTRY_RESPONSIBILITY,1
ICB,36
NHS_REGION,7
SUB_ICB_LOC,106


In [204]:
# Cross-check entity counts by dividing row counts by the 5 rate measure types.
unique_org_counts = (nhs_rate["ORG_TYPE"].value_counts() / 5).astype(int)

print("Unique entity count for geographic levels (organisations)")
unique_org_counts.to_frame(name="UNIQUE_ORG_COUNT")


Unique entity count for geographic levels (organisations)


,UNIQUE_ORG_COUNT
ORG_TYPE,
SUB_ICB_LOC,106
ICB,36
NHS_REGION,7
COUNTRY_RESPONSIBILITY,1


### MEASURE

In [205]:
# Check the five rate measures and how often each appears.
display(nhs_rate["MEASURE"].value_counts().to_frame())


,count
MEASURE,
DEMENTIA_ESTIMATE_65_PLUS,150
DEMENTIA_REGISTER_65_PLUS,150
DIAG_RATE_65_PLUS,150
DIAG_RATE_65_PLUS_LL,150
DIAG_RATE_65_PLUS_UL,150


In [206]:
# Confirm that every reporting entity has exactly 5 distinct rate measures.
nhs_rate.groupby(["ORG_TYPE", "ORG_CODE"])["MEASURE"].nunique().value_counts().to_frame(
    name="ENTITY_COUNT"
)


,ENTITY_COUNT
MEASURE,
5,150


Don't need to check the other fields as they will just show 150 unique items for the 5 measure types, as well as dates, as well as NaN for DQ

### Rate Dataset Structure Breakdown

The `rate` dataframe contains **750 rows** across **150 reporting entities**.

Each reporting entity has exactly **5 distinct measure records**, corresponding to the five diagnosis-rate measures in this release.

| Organisation Type (`ORG_TYPE`) | Total Rows | Unique Reporting Entities |
| :--- | ---: | ---: |
| **SUB_ICB_LOC** | 530 | **106** Sub-ICB locations |
| **ICB** | 180 | **36** Integrated Care Boards |
| **NHS_REGION** | 35 | **7** NHS regions |
| **COUNTRY_RESPONSIBILITY** | 5 | **1** country (`ENGLAND`) |
| **Total** | **750** | **150** |

The rate dataset therefore has a simple, regular grain:

`reporting entity + reporting date + measure = one value`


## Sub-ICB Dataset

**Publisher definition:** Primary care dementia data: Dementia data by sub ICB location

**Shape:** 11,660 rows × 13 columns

### What does one row represent?

The Sub-ICB dataset reports one measure/value for a Sub-ICB at the reporting period end date.

Some measures are further split into breakdowns such as age and gender, ethnicity, dementia type, residential type, prescribing, frailty or referrals.

The effective grain is therefore:

`Sub-ICB + reporting date + measure + applicable breakdown/category = one value`

Key fields:

- `ACH_DATE` - achieved date
    - represents the reporting period end date
    - this terminology is inherited from the Quality and Outcomes Framework (QOF) system
- `MEASURE` - identifies the main type of measure being reported
- `BREAKDOWN` - identifies the specific breakdown applied to the measure
    - e.g. `AGE_GENDER`, `ETHNICITY`, `DEMENTIA_TYPE`
    - some measures have no additional breakdown and contain `NaN`
- `AGE` - age category for the observation
    - `ALL` indicates the value is not restricted to a particular age category
- `GENDER` - gender category for the observation
    - `ALL` indicates the value is not restricted to a particular gender
- `ETHNICITY` - ethnicity category for the observation
    - `ALL` indicates the value is not restricted to a particular ethnicity category
- `DEMENTIA_TYPE` - dementia type category for the observation
    - `ALL` indicates the value is not restricted to a particular dementia type
- `RESIDENTIAL_TYPE` - residential category for the observation
    - `ALL` indicates the value is not restricted to a particular residential type
- `VALUE` - count of the measure for the applicable breakdown


### ORG / Sub-ICB entities

In [207]:
# Confirm the dataset only contains Sub-ICB rows and count the unique locations.
display(sub_icb["ORG_TYPE"].value_counts().to_frame())

display(
    sub_icb[["ODS_CODE", "ONS_CODE", "NAME"]]
    .nunique()
    .to_frame(name="UNIQUE_COUNT")
)


,count
ORG_TYPE,
SUB_ICB,11660


,UNIQUE_COUNT
ODS_CODE,106
ONS_CODE,106
NAME,106


In [208]:
# Check whether each Sub-ICB contributes the same number of observations.
sub_icb.groupby("ODS_CODE").size().value_counts().sort_index().to_frame(
    name="SUB_ICB_COUNT"
)


,SUB_ICB_COUNT
110,106


### MEASURE

In [209]:
# Count observations for each top-level dementia measure.
sub_icb["MEASURE"].value_counts().to_frame()


,count
MEASURE,
DEMENTIA_REGISTER,4876
PAT_LIST,2544
MCI,2332
PRESCRIBING,742
FRAILTY,530
REFERRALS,212
DELIRIUM_12M,106
INCIDENCE,106
PALLIATIVE_CARE,106


In [210]:
# Check that every Sub-ICB contains the same set of top-level measures.
sub_icb.groupby("ODS_CODE")["MEASURE"].nunique().value_counts().to_frame(
    name="SUB_ICB_COUNT"
)


,SUB_ICB_COUNT
MEASURE,
10,106


### BREAKDOWN

In [211]:
# Count the breakdown categories used across all Sub-ICB observations.
sub_icb["BREAKDOWN"].value_counts(dropna=False).to_frame()


,count
BREAKDOWN,
AGE_GENDER,7420
DEMENTIA_TYPE,848
ETHNICITY,848
RESIDENCE_TYPE,636
NaN,424
DEM_FALL_12M,106
DEM_FRAC_NOF_12M,106
DEM_MILD_FRAILTY,106
DEM_MOD_FRAILTY,106


In [212]:
# See which breakdowns belong to each top-level measure.
sub_icb[
    ["MEASURE", "BREAKDOWN"]
].drop_duplicates().sort_values(
    ["MEASURE", "BREAKDOWN"],
    na_position="first",
)


,MEASURE,BREAKDOWN
0,DELIRIUM_12M,NaN
1,DEMENTIA_REGISTER,AGE_GENDER
25,DEMENTIA_REGISTER,DEMENTIA_TYPE
33,DEMENTIA_REGISTER,ETHNICITY
41,DEMENTIA_REGISTER,RESIDENCE_TYPE
47,FRAILTY,DEM_FALL_12M
48,FRAILTY,DEM_FRAC_NOF_12M
49,FRAILTY,DEM_MILD_FRAILTY
50,FRAILTY,DEM_MOD_FRAILTY
51,FRAILTY,DEM_SEV_FRAILTY


### AGE / GENDER

In [213]:
# Inspect the age categories used where measures are split by age.
sub_icb["AGE"].value_counts().to_frame()


,count
AGE,
ALL,4240
40_44,636
45_49,636
50_54,636
55_59,636
60_64,636
65_69,636
70_74,636
75_79,636


In [214]:
# Inspect the gender categories used where measures are split by gender.
sub_icb["GENDER"].value_counts().to_frame()


,count
GENDER,
ALL,4240
Female,2438
Male,2438
FEMALE,1272
MALE,1272


### ETHNICITY

In [215]:
# Inspect the ethnicity categories represented in the Sub-ICB data.
sub_icb["ETHNICITY"].value_counts().to_frame()


,count
ETHNICITY,
ALL,10812
ASIAN_OR_ASIAN_BRITISH,106
BLACK_OR_AFRICAN_OR_CARIBBEAN_OR_BLACK_BRITISH,106
INCONCLUSIVE_ETHNIC_GROUP,106
MIXED_OR_MULTIPLE_ETHNIC_GROUPS,106
NOT_DEFINED,106
NOT_STATED,106
OTHER_ETHNIC_GROUP,106
WHITE,106


### DEMENTIA TYPE

In [216]:
# Inspect the dementia type categories represented in the Sub-ICB data.
sub_icb["DEMENTIA_TYPE"].value_counts().to_frame()


,count
DEMENTIA_TYPE,
ALL,10812
ALZHEIMERS_DISEASE,106
FRONTOTEMPORAL,106
INCONCLUSIVE_DEMENTIA_TYPE,106
LEWY_BODY,106
MIXED_DEMENTIA_TYPES,106
OTHER_SPECIFIED_DEMENTIA_TYPES,106
OTHER_UNSPECIFIED_DEMENTIA_TYPES,106
VASCULAR_DEMENTIA,106


### RESIDENTIAL TYPE

In [217]:
# Inspect the residential type categories represented in the Sub-ICB data.
sub_icb["RESIDENTIAL_TYPE"].value_counts().to_frame()


,count
RESIDENTIAL_TYPE,
ALL,11024
INCONCLUSIVE_RESIDENTIAL_TYPE,106
NO_PERMANENT_ADDRESS,106
NURSING_HOME,106
OTHER_RESIDENTIAL_TYPE,106
PRIVATE_RESIDENCE,106
RESIDENTIAL_CARE_HOME,106


### Sub-ICB Dataset Structure Breakdown

The dataset contains **11,660 observations** across **106 Sub-ICB locations**.

In this release, every Sub-ICB contributes exactly **110 rows** and contains the same **10 top-level measures**. The 110 rows arise because some measures have one overall value while others expand into multiple breakdown/category rows.

| Measure | Rows per Sub-ICB | Structure |
| :--- | ---: | :--- |
| `DEMENTIA_REGISTER` | 46 | age/gender, ethnicity, dementia type and residence type |
| `PAT_LIST` | 24 | age/gender |
| `MCI` | 22 | age/gender |
| `PRESCRIBING` | 7 | seven prescribing breakdowns |
| `FRAILTY` | 5 | five frailty breakdowns |
| `REFERRALS` | 2 | two referral breakdowns |
| `DELIRIUM_12M` | 1 | overall value |
| `INCIDENCE` | 1 | overall value |
| `PALLIATIVE_CARE` | 1 | overall value |
| `YOUNG_ONSET` | 1 | overall value |

Unlike the rate dataset, the different rows for a Sub-ICB are not simply repeated versions of one fixed measure list. The `BREAKDOWN` and category columns form part of the observation grain.


## Practice Dataset

**Publisher definition:** Primary care dementia data: Recorded diagnoses, care plans, reviews by GP practice

**Shape:** 42,616 rows × 7 columns

### What does one row represent?

The practice dataset reports one measure/breakdown value for a GP practice at the reporting period end date.

The effective grain is:

`GP practice + reporting date + measure + breakdown = one value`

Key fields:

- `ACH_DATE` - reporting period end date
- `ORG_TYPE` - organisational breakdown
    - all rows in this file use `PRACTICE`
- `ODS_CODE` - Organisation Data Service identifier for the GP practice
- `NAME` - name of the GP practice
- `MEASURE` - groups related data items into a top-level measure
- `BREAKDOWN` - identifies the specific data item within the measure
- `VALUE` - count of the measure/breakdown


### ORG / GP practices

In [218]:
# Confirm the organisation type and count the unique GP practice codes.
display(practice["ORG_TYPE"].value_counts().to_frame())

print(f"Unique GP practice codes: {practice['ODS_CODE'].nunique():,}")


,count
ORG_TYPE,
PRACTICE,42616


Unique GP practice codes: 6,088


In [219]:
# Check whether every GP practice contributes the same number of observations.
practice.groupby("ODS_CODE").size().value_counts().sort_index().to_frame(
    name="PRACTICE_COUNT"
)


,PRACTICE_COUNT
7,6088


### MEASURE

In [220]:
# Count observations for each top-level practice measure.
practice["MEASURE"].value_counts().to_frame()


,count
MEASURE,
REVIEWS,18264
DEMENTIA_REGISTER,12176
PAT_LIST,12176


### BREAKDOWN

In [221]:
# Count the individual breakdown/data items reported for practices.
practice["BREAKDOWN"].value_counts().to_frame()


,count
BREAKDOWN,
DEMENTIA_REGISTER_0_64,6088
DEMENTIA_REGISTER_65_PLUS,6088
PAT_LIST_0_64,6088
PAT_LIST_65_PLUS,6088
DIAG_DECLINED_CARE_PLAN,6088
DIAG_RECEIVED_CARE_PLAN,6088
DIAG_RECEIVED_MED_REV,6088


In [222]:
# See how each practice breakdown maps to its top-level measure.
practice[
    ["MEASURE", "BREAKDOWN"]
].drop_duplicates().sort_values(
    ["MEASURE", "BREAKDOWN"]
)


,MEASURE,BREAKDOWN
0,DEMENTIA_REGISTER,DEMENTIA_REGISTER_0_64
1,DEMENTIA_REGISTER,DEMENTIA_REGISTER_65_PLUS
2,PAT_LIST,PAT_LIST_0_64
3,PAT_LIST,PAT_LIST_65_PLUS
4,REVIEWS,DIAG_DECLINED_CARE_PLAN
5,REVIEWS,DIAG_RECEIVED_CARE_PLAN
6,REVIEWS,DIAG_RECEIVED_MED_REV


### Practice Dataset Structure Breakdown

The dataset contains **42,616 observations** across **6,088 GP practice codes**.

Every practice contributes exactly **7 rows** in this release:

| Measure | Rows per Practice | Breakdown |
| :--- | ---: | :--- |
| `DEMENTIA_REGISTER` | 2 | aged 0-64; aged 65+ |
| `PAT_LIST` | 2 | aged 0-64; aged 65+ |
| `REVIEWS` | 3 | declined care plan; received care plan; received medication review |
| **Total** | **7** | |

This gives the practice dataset a regular grain of one row per practice, reporting date and measure/breakdown combination.

**Dictionary note:** the data dictionary lists the `PAT_LIST_65_PLUS` row with `MEASURE = PAT_LIST_65_PLUS`, while the CSV itself uses `MEASURE = PAT_LIST` with `BREAKDOWN = PAT_LIST_65_PLUS`. The observed CSV structure is used here, with the discrepancy noted for later validation.


## Mapping Dataset

**Publisher definition:** Primary care dementia data: current month: Mapping (Commissioning Regions-ICBs-SICBLs-GP practice)

**Shape:** 6,182 rows × 13 columns

### What does one row represent?

The mapping dataset is a snapshot linking one GP practice to the NHS geographies/organisations above it at the extraction date.

The effective grain is:

`GP practice at extract date = one mapping row`

Key fields:

- `EXTRACT_DATE` - date the mapping snapshot was extracted
- `PRACTICE_CODE` - GP practice code
- `PRACTICE_NAME` - GP practice name
- `PRACTICE_POSTCODE` - GP practice postcode
- `ONS_SUB_ICB_LOCATION_CODE` - ONS 9-digit geography code for the Sub-ICB location
- `SUB_ICB_LOCATION_CODE` - ODS code for the Sub-ICB location
- `SUB_ICB_LOCATION_NAME` - Sub-ICB location name
- `ONS_ICB_CODE` - ONS 9-digit geography code for the ICB
- `ICB_CODE` - ODS code for the ICB
- `ICB_NAME` - ICB name
- `ONS_COMM_REGION_CODE` - ONS 9-digit geography code for the NHS commissioning region
- `COMM_REGION_CODE` - ODS code for the NHS commissioning region
- `COMM_REGION_NAME` - NHS commissioning region name


### PRACTICE / GEOGRAPHY COVERAGE

In [223]:
# Confirm the mapping is one row per practice code and inspect the extract date.
print(f"Rows: {len(mapping):,}")
print(f"Unique practice codes: {mapping['PRACTICE_CODE'].nunique():,}")

display(mapping["EXTRACT_DATE"].value_counts().to_frame())


Rows: 6,182
Unique practice codes: 6,182


,count
EXTRACT_DATE,
01-Jul-26,6182


In [224]:
# Count the unique geographic organisations represented by the mapping.
mapping[
    ["SUB_ICB_LOCATION_CODE", "ICB_CODE", "COMM_REGION_CODE"]
].nunique().to_frame(name="UNIQUE_COUNT")


,UNIQUE_COUNT
SUB_ICB_LOCATION_CODE,106
ICB_CODE,36
COMM_REGION_CODE,7


In [225]:
# Check whether any practice rows are missing their geographic mapping.
mapping[
    [
        "SUB_ICB_LOCATION_CODE",
        "ICB_CODE",
        "COMM_REGION_CODE",
    ]
].isna().sum().to_frame(name="MISSING_COUNT")


,MISSING_COUNT
SUB_ICB_LOCATION_CODE,1
ICB_CODE,1
COMM_REGION_CODE,1


### PRACTICE DATASET COMPARISON

In [226]:
# Compare GP practice coverage between the practice observations and mapping snapshot.
practice_codes = set(practice["ODS_CODE"])
mapping_codes = set(mapping["PRACTICE_CODE"])

print(f"Practice dataset codes: {len(practice_codes):,}")
print(f"Mapping dataset codes: {len(mapping_codes):,}")
print(f"Present in both: {len(practice_codes & mapping_codes):,}")
print(f"Mapping only: {len(mapping_codes - practice_codes):,}")
print(f"Practice data only: {len(practice_codes - mapping_codes):,}")


Practice dataset codes: 6,088
Mapping dataset codes: 6,182
Present in both: 6,088
Mapping only: 94
Practice data only: 0


### Mapping Dataset Structure Breakdown

The mapping file contains **6,182 rows** and **6,182 unique GP practice codes**, confirming one mapping row per practice code in this snapshot.

The hierarchy represented is:

`GP practice → Sub-ICB location → ICB → NHS commissioning region`

The mapping contains:

- **106** Sub-ICB location codes
- **36** ICB codes
- **7** NHS commissioning region codes

There is **1 practice row** with missing Sub-ICB, ICB and commissioning-region mapping fields.

Comparing the mapping snapshot with the practice observations:

- all **6,088** practice codes in the practice dataset are present in the mapping file
- the mapping file contains an additional **94** practice codes that do not appear in the June practice observations

The current files establish that the mapping snapshot covers more practice codes than the practice dataset; the reason for those additional 94 practices is not established at this stage.


## What this structure pass has established

At this stage the aim has not been to analyse dementia outcomes. It has been to understand the **grain and structure** of the source files before analysing values.

So far this has established:

- what one row means in each dataset
- which columns identify organisations, measures and breakdowns
- which datasets have a fixed/repeated structure across organisations
- how the different measures expand into dimensional breakdowns
- how GP practices map into Sub-ICBs, ICBs and NHS regions
- where the source files contain small inconsistencies or missing mappings worth remembering

This gives a basis for deciding later how the raw NHS files should be validated, transformed and modelled for the Atlas.
